src/pal/pal_device.cpp

```diff
@@ -0,0 +1,541 @@
+
+    result = device_->CreateQueue(queue_create_info, malloc(queue_size), &queue);
+    CHECK_PAL(result, "CreateQueue()");
+
+    // Create the command buffer.
+    Pal::CmdBufferCreateInfo cmd_buffer_create_info = {};
+    cmd_buffer_create_info.pCmdAllocator = cmd_allocator_;
+    cmd_buffer_create_info.queueType = queue_type;
+    cmd_buffer_create_info.engineType = engine_type;
+
+    const size_t cmdBufSize = device_->GetCmdBufferSize(cmd_buffer_create_info, &result);
+    CHECK_PAL(result, "GetCmdBufferSize()");
+
+    return device_->CreateCmdBuffer(cmd_buffer_create_info, malloc(cmdBufSize), &cmd_buffer);
+}
+
+PalDevice::GpuVirtAddr_t PalDevice::allocate_gpu_memory(Pal::gpusize size_in_bytes, Pal::GpuHeap heap) {
+    Pal::IGpuMemory* memory;
+    Pal::Result result = allocate_memory(size_in_bytes, heap, &memory);
+    CHECK_PAL(result, "allocate_gpu_memory(GpuHeapInvisible)");
+    const auto gpu_virtual_addr = track_memory(memory);
+    return gpu_virtual_addr;
+}
+
+Pal::Result PalDevice::allocate_memory(
+    Pal::gpusize size_in_bytes, Pal::GpuHeap heap, Pal::IGpuMemory** gpu_memory_pp, Pal::gpusize alignment) {
+    Pal::GpuMemoryHeapProperties heap_properties[Pal::GpuHeapCount] = {};
+    Pal::Result result = device_->GetGpuMemoryHeapProperties(&heap_properties[0]);
+    CHECK_PAL(result, "GetGpuMemoryHeapProperties()");
+
+    Pal::GpuMemoryCreateInfo create_info = {};
+    create_info.alignment = alignment;
+    create_info.size = size_in_bytes;
+    create_info.flags.cpuInvisible = ((heap_properties[heap].flags.cpuVisible == 0) ? 1 : 0);
+    create_info.vaRange = Pal::VaRange::Default;
+    create_info.heapCount = 1;
+    create_info.priority = Pal::GpuMemPriority::Normal;
+    create_info.priorityOffset = Pal::GpuMemPriorityOffset::Offset0;
+    create_info.heaps[0] = heap;
+    create_info.mallPolicy = Pal::GpuMemMallPolicy::Always;
+
+    const size_t gpu_memory_size = device_->GetGpuMemorySize(create_info, &result);
+    CHECK_PAL(result, "GetGpuMemorySize()");
+
+    result = device_->CreateGpuMemory(create_info, malloc(gpu_memory_size), gpu_memory_pp);
+    CHECK_PAL(result, "CreateGpuMemory()");
+
+    Pal::GpuMemoryRef mem_ref = {};
+    mem_ref.pGpuMemory = *gpu_memory_pp;
+
+    // Make this allocation permanently resident. We might want to track  which allocations are actually being
+    // used and only make those resident but in AnyDSL usually the user takes care of releasing memory as soon
+    // as it's not used anymore.
+    return device_->AddGpuMemoryReferences(1, &mem_ref, nullptr, Pal::GpuMemoryRefCantTrim);
+}
+
+PalDevice::GpuVirtAddr_t PalDevice::allocate_shared_virtual_memory(Pal::gpusize sizeInBytes) {
+    Pal::SvmGpuMemoryCreateInfo create_info = {};
+    create_info.size = sizeInBytes;
+    create_info.alignment =
+        4096; // This seems to be the minimum alignment needed to avoid triggering asserts in PAL
+    create_info.mallPolicy = Pal::GpuMemMallPolicy::Always;
```

Comment:
for my curiosity: what does "mall policy" mean ?

Reply:
MALL is the last level cache on AMD GPUs. You can define for each allocation whether it should bypass the last-level cache or not.